lib/usb/usb_process.js

```diff
@@ -1,236 +1,271 @@
+  // 4 remote process streams
+  this.control = new RemoteWritableStream(id, daemon, protocol.controlWrite, protocol.controlClose);
+  this.stdin = new RemoteWritableStream(id, daemon, protocol.stdinWrite, protocol.stdinClose);
+  this.stdout = new RemoteReadableStream(id, daemon, protocol.stdoutAck);
+  this.stderr = new RemoteReadableStream(id, daemon, protocol.stderrAck);
 
   // Once the process is killed
   self.once('death', function(data) {
-    self.active = false;
-    // Set the exit code
-    self.exitCode = data.arg;
-    // Close the input streams
+    // Close the input streams because the remote proc can't handle data
     self.control.end();
     self.stdin.end();
-    // Clean up remote resources
-    self.close();
+
+    // Mark this process as inactive
+    self.active = false;
   });
 
-  this.close = function() {
+  // Tell the USBDaemon to close the remote process
+  self.close = function() {
     // Write the close command
     daemon.write(protocol.closeProcess(self.id));
-    self.once('close', function() {
-
-      // End the output streams
-      self.stdout.push(null);
-      self.stderr.push(null);
-    })
   };
 
-  this.kill = function(signal) {
+  // Send a signal to the remote process
+  self.kill = function(signal) {
+    // Write the signal command
     daemon.write(protocol.killProcess(self.id, signal));
   }
 
-
-  this.control.on('drain', function() {
-    self._drainBackPressure(self.control, protocol.controlWrite);
-  });
-
-
-  this.stdin.on('drain', function() {
-    self._drainBackPressure(self.stdin, protocol.stdinWrite);
-  });
-
-  this.control._write = function(chunk, enc, callback) {
-    // Add this chunk to the backpressure array
-    self.control.backPressure.push({buffer: chunk, callback: function controlSent() {
-      // Tell the daemon that we have finished writing to the control stream
-      daemon.write(protocol.controlClose(self.id));
-      // Call the callback
-      callback && callback();
-    }});
-
-    if (self.control.credit) {
-      self._drainBackPressure(self.control, protocol.controlWrite);
+  // Checks if this process is ready to be closed
+  // and closes it if it can
+  self.closeProcessIfReady = function() {
+    // If this process hasn't been closed before
+    // and all of the streams are closed
+    if (self.active && self.stdout.closed
+            && self.stderr.closed
+            && self.stdin.closed
+            && self.control.closed) {
+      // Close this process
+      self.close();
     }
   };
 
-  // Intercepting write to the stdin stream
-  this.stdin._write = function(chunk, enc, callback) {
-    // Add this chunk to the backpressure array
-    self.stdin.backPressure.push({buffer: chunk, callback: callback});
-    if (self.stdin.credit) {
-      self._drainBackPressure(self.stdin, protocol.stdinWrite);
-    }
-  };
+  // When any of the streams complete, check if the process is ready to close
+  self.stdin.once('finish', self.closeProcessIfReady);
```

Comment:
If you remove stdin from the condition as described above, you can also remove this.